### ISSUE TITLE
Equipping into an NFT with 2 resources of same base

### ISSUE THREAD

**Issue by Swader** [rmrk-team/evm#18]
<img width="555" alt="Screenshot 2022-03-31 at 14 47 51" src="https://user-images.githubusercontent.com/1430603/161058154-4a9971a5-716f-4a5e-b645-bc418338c02a.png">

We should perhaps modify the Equip logic such that it pays attention to which resource is is equipping into as well.

**Comment by steven2308** [comment]
I will assume a "base instance" in this context is a resource pointing to a base. (There are no such things as base instances, bases are just for reference).
In the current implementation this is not possible and I don't think it should be. We allow one resource to point to a single base and a single slot. **This makes it easier on code and also avoids unexpected behavior, like a weapon resource meant for the left hand being equipped on the right hand or showing 2 weapons in a single hand.**

We keep track from the parent context  (say a bird) that only one NFT is equipped into each token/base/slot combination. We also keep track from the child context (say a hat) that it is only equipped into one NFT at a time (this is important for scarcity). 

Just FYI, we also check on the base, that the child which wants to be equipped is meant to be in that slot for the base, this is to avoid bad use of the base, only the issuer can declare what collections can be used for the base (or if a slot part is equippable by any collection).

Acout also proposed an interesting example of a soldier loosing a hand after a war, so the hand shouldn't be able to equip weapons anymore. In this case, the solution would be to send a new resource to the token (possibly replacing the original one). This new resource would use a new Base, which does not have a slot for the lost hand.

@Swader  what are your thoughts on this?

**Comment by Ayuilos** [comment]
In fact, what I want to say will also not cause the situation mentioned in steven's bold text. I think there is a concept of Base instance, because when we use Base, we do not directly quote but cherry-pick, then every resource that uses Base is an instance of Base.
Therefore, it is reasonable to have multiple instances referencing the same Base in an NFT.

For more context please let steven, me or cicada invite you to our discussion group.

I think there's a need to reach a consensus.

**Comment by Ayuilos** [comment]
> I will assume a "base instance" in this context is a resource pointing to a base. (There are no such things as base instances, bases are just for reference). In the current implementation this is not possible and I don't think it should be. We allow one resource to point to a single base and a single slot. **This makes it easier on code and also avoids unexpected behavior, like a weapon resource meant for the left hand being equipped on the right hand or showing 2 weapons in a single hand.**
> 
> We keep track from the parent context (say a bird) that only one NFT is equipped into each token/base/slot combination. We also keep track from the child context (say a hat) that it is only equipped into one NFT at a time (this is important for scarcity).
> 
> Just FYI, we also check on the base, that the child which wants to be equipped is meant to be in that slot for the base, this is to avoid bad use of the base, only the issuer can declare what collections can be used for the base (or if a slot part is equippable by any collection).
> 
> Acout also proposed an interesting example of a soldier loosing a hand after a war, so the hand shouldn't be able to equip weapons anymore. In this case, the solution would be to send a new resource to the token (possibly replacing the original one). This new resource would use a new Base, which does not have a slot for the lost hand.
> 
> @Swader what are your thoughts on this?

I think the core problem is here:
> In this case, the solution would be to send a new resource to the token (possibly replacing the original one). This new resource would use a new Base, which does not have a slot for the lost hand.
We don't need this slot by not cherry-picking this slot is enough, we still use the same Base, maybe there's no need to add a new Base.

**Comment by steven2308** [comment]
Slots parts cannot be cherry picked. When a resource uses a base (and it can only use one), it can receive NFTs in all the slots that the base defines.
You can only cherry pick fixed parts for a resource. For instance, if body parts are defined in the base, you're new resource could simply not pick a fixed part for the arm, but unless it uses a different base you would still be able to equip a weapon into a missing arm. So you either use a different base without the slot, or wrap the equipping logic to control that on your contract.

This functionality of cherry picking slot parts is not in the spec AFAIR, it would make everything much more complex, and there's a simple alternative of just using a new resource with a different base.

**Comment by Ayuilos** [comment]
Any example for pointing that using the asset ID for equipments mapping would be a huge issue?

**Comment by steven2308** [comment]
> just needing to replay the equip action on new asset.

This is not a "just", it's expensive an bad UX.
For historical data, we recommend indexers always.

**Comment by steven2308** [comment]
So just to be clear, you're suggestion is to map through asset id, right?

**Comment by Ayuilos** [comment]
Yes because I think only in this way is respecting ERC-5773 philosophy.
The situation you mentioned above is adding custom logic to auto replace equipments, rather than breaking ERC-5773 definition.

**Comment by steven2308** [comment]
I don't understand why you consider this would not be respecting 5773, could you elaborate? What are we breaking from it's definition?

**Comment by Ayuilos** [comment]
OK

We know that the definition of the ERC-5773 is that the NFT produces different outputs based on its context. Each asset, in fact, reflects the output of the NFT in some context. This is actually a philosophical description of "There are a thousand Hamlets in a thousand people's eyes." What we see is the projection of the NFT into our perspective, and the NFT itself is an abstraction.

My understanding of the equipment properties, then, is that the equipment properties allow NFT A and NFT B to enter the same context and produce corresponding chemical reaction projections based on the ownership relationship brought about by the ERC-6059, so the ERC-6220 is a special representation of the ERC-5773.

If you agree with me above, then the equiping behavior should occur on the assets, because each asset should correspond to a separate context. This is the foundation of ERC-5773, and now ERC-6220 says that all assets referencing the same Catalog should correspond to the same context. This breaks the most fundamental definition of ERC-5773.

**Comment by Ayuilos** [comment]
Let's take an example where we add two ordinary assets but the metadata of both assets may point to the same URI for different reasons, can you assume that they are the same asset just because they have the same metadata?

If a light shines on an object, the light shines on the left or the light shines on the right, we may see the same projection on the wall, but the shines action is different.
